codegen/type_converter_test.go

```diff
@@ -966,3 +1026,542 @@ func TestConvertWithBadKeyMapOfString(t *testing.T) {
+	fieldMap["One"] = codegen.FieldMapperEntry{
+		QualifiedName: "Two",
+		Override:      false,
+	}
+
+	lines, err := convertTypes(
+		"Foo", "Bar",
+		`struct Foo {
+			1: optional bool one
+			2: optional bool two
+		}
+
+		struct Bar {
+			1: optional bool one
+			2: optional bool two
+		}`,
+		nil,
+		fieldMap,
+	)
+
+	assert.NoError(t, err)
+	assert.Equal(t, trim(`
+		out.One = (*bool)(in.Two)
+		if in.One != nil {
+			out.One = (*bool)(in.One)
+		}
+		out.Two = (*bool)(in.Two)
+	`), lines)
+}
+
+func TestConverterMapOverrideOptional(t *testing.T) {
+	fieldMap := make(map[string]codegen.FieldMapperEntry)
+	fieldMap["One"] = codegen.FieldMapperEntry{
+		QualifiedName: "Two",
+		Override:      true,
+	} // Map from required filed, unconditional assignment
+	fieldMap["Two"] = codegen.FieldMapperEntry{
+		QualifiedName: "One",
+		Override:      true,
+	}
+	fieldMap["Three"] = codegen.FieldMapperEntry{
+		QualifiedName: "Four",
+		Override:      true,
+	}
+	fieldMap["Four"] = codegen.FieldMapperEntry{
+		QualifiedName: "Three",
+		Override:      false,
+	} // Map to required filed, unconditional assignment
+
+	lines, err := convertTypes(
+		"Foo", "Bar",
+		`struct Foo {
+			1: optional bool one
+			2: optional bool two
+			3: optional bool three
+			4: required bool four
+		}
+
+		struct Bar {
+			1: optional bool one
+			2: required bool two
+			3: optional bool three
+			4: optional bool four
+		}`,
+		nil,
+		fieldMap,
+	)
+
+	assert.NoError(t, err)
+	assert.Equal(t, trim(`
+		out.One = (*bool)(in.Two)
+		out.Two = bool(in.Two)
+		if in.One != nil {
+			out.Two = bool(in.One)
+		}
+		out.Three = (*bool)(in.Three)
+		if in.Four != nil {
+			out.Three = (*bool)(in.Four)
+		}
+		out.Four = (*bool)(in.Four)
+		`), lines)
+}
+
+func TestConverterMapStructWithSubFields(t *testing.T) {
+	fieldMap := make(map[string]codegen.FieldMapperEntry)
+	fieldMap["Three.One"] = codegen.FieldMapperEntry{
+		QualifiedName: "Three.Two",
+		Override:      true,
+	}
+	fieldMap["Four.Two"] = codegen.FieldMapperEntry{
+		QualifiedName: "Four.One",
+		Override:      true,
+	}
+
+	lines, err := convertTypes(
+		"Foo", "Bar",
+		`struct NestedFoo {
+			1: required string one
+			2: optional string two
+		}
+
+		struct NestedBar {
+			1: required string one
+			2: optional string two
+		}
+
+		struct Foo {
+			3: optional NestedFoo three
+			4: required NestedFoo four
+		}
+
+		struct Bar {
+			3: optional NestedBar three
+			4: required NestedBar four
+		}`,
+		nil,
+		fieldMap,
+	)
+
+	assert.NoError(t, err)
+	assert.Equal(t, trim(`
+		if in.Three != nil {
+			out.Three = &structs.NestedBar{}
+			out.Three.One = string(in.Three.One)
+			if in.Three.Two != nil {
+				out.Three.One = string(in.Three.Two)
+			}
+			out.Three.Two = (*string)(in.Three.Two)
+		} else {
+			out.Three = nil
+		}
+		if in.Four != nil {
+			out.Four = &structs.NestedBar{}
+			out.Four.One = string(in.Four.One)
+			out.Four.Two = (*string)(in.Four.One)
```

Comment:
cannot cast string to `*string` generated code does not work.

Need to use `ptr.String(in.Four.One)` to cast from required to optional.